src/monogame/common/MGM.cpp

```diff
@@ -0,0 +1,92 @@
+// MonoGame - Copyright (C) The MonoGame Team
+// This file is subject to the terms and conditions defined in
+// file 'LICENSE.txt', which is part of this source code package.
+
+#include "csharp_MGM.h"
+
+#include "stl_common.h"
+
+
+struct MGM_Song
+{
+	mgulong duration;
+
+	void (*finishCallback)(void*);
+	void* finishData;
+};
+
+MGM_Song* MGM_Song_Create(const char* mediaFilePath)
+{
+	assert(mediaFilePath != nullptr);
+
+	// TODO: This should detect the media format using
+	// standard native format parsing libraries that are
+	// portable to all our target platforms:
+	//
+	// libvorbis
+	// minimp3
+	// wave ??
+	// mp4 ??
+	// FLAC ??
+	//
+	// It should then spin up thread which decodes the
+	// audio and streams buffers to a native SoundEffect.
```

Comment:
@mrhelmut @harry-cpp 

So the idea here is Song can be shared code across all platforms.

It takes care of streaming a music file from disk and pushing decoded packets into the platform specific sound effect.

Reply:
Songs with some platforms are played via SoundEffects, however where possible they are played via the Platform's built-in media player and passed by URI. 

My understanding is this should provide integration (where possible) with not playing over other music apps (MediaPlayer.GameHasControl), playing/shuffling music from the user's own music library, allowing playback of DRM protected files in that library, getting album artwork, audio visualization data, etc.

Quite often I have seen developer's ask how to cross fade between 2 Songs or play additional layer tracks on top of each other. Song isn't really designed for this (one song for the device at any time) whereas SoundEffect fits this purpose well. At a later point it might be useful to have a PlatformBehavior setting for playback preference (by device media player or SoundEffects).

Reply:
> where possible they are played via the Platform's built-in media player and passed by URI.
> My understanding is this should provide integration (where possible)

@squarebananas  this is the problem.

"Where possible" is just two platforms.... Android and iOS.  No other console or platform has these concepts of unified music libraries.  It made sense in 2004... it makes no sense in 2024.  Is it going to integrate with Spotify?

The point is...  almost no games need these features.  Ones that do can code to Android/iOS/Spotify APIs directly.

99% of MG/XNA users just want to efficiently stream a song from content/files in the game. 

> Quite often I have seen developer's ask how to cross fade between 2 Songs or play additional layer tracks on top of each other.

I would like to see us support this functionality in the future, and we can if we don't tie ourselves to old platform media players.

This implementation is designed for that future.

> whereas SoundEffect fits this purpose well.

If you want to manually deal with streaming or punt and load the whole sound into memory.  Sure.

Most devs don't want to mess with streaming sounds.  Song i feel fits that roll.

> At a later point it might be useful to have a PlatformBehavior setting for playback preference (by device media player or SoundEffects).

I think this is would be a waste of MG development resources.  I've seen maybe 2 or 3 MG/XNA games in the last 10 years that wanted to integrate with the device media player... mostly music rhythm games.

Every other game just wants to stream music files.

Reply:
Ok thanks, that all makes sense. I didn't realize newer Xbox consoles no longer have those features (used to use the Xbox 360 audio player a lot).

If the plan is to remove platform media players then there won't be any issues. Mainly I was just worried you might develop new shared Song code and then realize not all platforms currently use SoundEffects for this.

When I said "SoundEffect fits this purpose well" I was thinking in terms of the current implementations of the SoundEffect/SoundEffectInstance/DynamicSoundInstance system vs Song/MediaPlayer system. Making Song a simplified version of DynamicSoundInstance where it manages BufferNeeded events for the user sounds like a good idea.